### PR TITLE
Add support for permission offer/response

### PR DESCRIPTION
**Pull request by austinzheng**
Changes:
- Added APIs for creating permission offers and accepting them
- Refactored some code to facilitate reuse
- Renamed a few types to reflect their new, expanded usage

Rationale: right now, most bindings (Cocoa included) use the object store APIs to get and change permissions, but binding-level APIs to create and accept permission offers. This API allows the existing object store machinery to be used to perform the latter task as well.

More detailed description.
1. I broke most of the existing method for changing permissions into a new method, `perform_async_operation()`. This method basically creates an object, writes it to the management Realm, and observes it until its status code indicates either success or failure. Then it calls a callback.
2. Changing permissions, creating offers, and accepting offers now call into `perform_async_operation()` with customization options (e.g. what object to create, what additional properties to set on it).

### CODE REVIEW

File: src/sync/sync_permission.cpp [hunk 1]
```diff
@@ -157,7 +157,91 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
...
 {
-    const auto realm_url = user->server_url() + permission.path;
+    auto props = AnyDict{
+        {"userId", permission.condition.user_id},
+        {"realmUrl", user->server_url() + permission.path},
+        {"mayRead", permission.access != AccessLevel::None},
+        {"mayWrite", permission.access == AccessLevel::Write || permission.access == AccessLevel::Admin},
+        {"mayManage", permission.access == AccessLevel::Admin},
+    };
+    switch (permission.condition.type) {
```
Comment (tgoyne) [inline src/sync/sync_permission.cpp]: A one-case switch is sort of silly.

File: src/sync/sync_permission.cpp [hunk 2]
```diff
@@ -157,7 +157,91 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
...
+        case Permission::Condition::Type::KeyValue:
+            props.insert({"metadataKey", permission.condition.key_value.first});
+            props.insert({"metadataValue", permission.condition.key_value.second});
+            break;
+        default:
+            break;
+    }
+    auto cb = [callback=std::move(callback)](Object*, std::exception_ptr exception) {
+        if (exception) {
```
Comment (tgoyne) [inline src/sync/sync_permission.cpp]: This if/else is redundant.

File: src/sync/sync_permission.cpp [hunk 3]
```diff
@@ -157,7 +157,91 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
...
+{
+    auto props = AnyDict{
+        {"token", token},
+    };
+    auto cb = [callback=std::move(callback)](Object* object, std::exception_ptr exception) {
+        if (exception) {
+            callback(none, exception);
+        } else {
+            CppContext context;
+            auto token = any_cast<std::string>(object->get_property_value<util::Any>(context, "realmUrl"));
```
Comment (tgoyne) [inline src/sync/sync_permission.cpp]: The string here is the only thing that differs from the above, so it's probably worth extracting this all to a function.

File: src/sync/sync_permission.cpp [hunk 4]
```diff
@@ -166,38 +250,27 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
...
-    // Set condition properties based on type
-    switch (permission.condition.type) {
-        case Permission::Condition::Type::KeyValue:
-            raw.set_property_value<util::Any>(context, "metadataKey", permission.condition.key_value.first, false);
-            raw.set_property_value<util::Any>(context, "metadataValue", permission.condition.key_value.second, false);
-            break;
-        default:
-            break;
+    };
+    for (auto&& prop : additional_props) {
```
Comment (tgoyne) [inline src/sync/sync_permission.cpp]: `props.insert(additional_props.begin(), additional_props.end());`

File: src/sync/sync_permission.cpp [hunk 5]
```diff
@@ -261,23 +325,23 @@ SharedRealm Permissions::management_realm(std::shared_ptr<SyncUser> user, const
...
             Property{"updatedAt",         PropertyType::Date},
-            Property{"expiresAt",         PropertyType::Date},
             Property{"statusCode",        PropertyType::Int|PropertyType::Nullable},
             Property{"statusMessage",     PropertyType::String|PropertyType::Nullable},
             Property{"token",             PropertyType::String|PropertyType::Nullable},
             Property{"realmUrl",          PropertyType::String},
             Property{"mayRead",           PropertyType::Bool},
             Property{"mayWrite",          PropertyType::Bool},
             Property{"mayManage",         PropertyType::Bool},
+            Property{"expiresAt",         PropertyType::Date|PropertyType::Nullable},
```
Comment (tgoyne) [inline src/sync/sync_permission.cpp]: Why is this moved to the end?

File: src/sync/sync_permission.hpp [hunk 6]
```diff
@@ -22,28 +22,31 @@
 #include "results.hpp"
 #include "shared_realm.hpp"
 
+#include "util/any.hpp"
```
Comment (tgoyne) [inline src/sync/sync_permission.hpp]: A forward declaration of `Any` suffices here.

File: src/sync/sync_permission.cpp (hunk 5 above)
Comment (austinzheng) [reply src/sync/sync_permission.cpp]: I was trying to troubleshoot something; forgot to move it back. Thank you for catching this.

File: src/sync/sync_permission.cpp [hunk 7]
```diff
@@ -213,28 +278,19 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
...
-                                  : util::format("Error code: %1", code);
-            callback(std::make_exception_ptr(PermissionChangeException(error_str, code)));
+            std::string error_str = (status.has_value()
+                                     ? any_cast<std::string>(status)
+                                     : util::format("Error code: %1", code));
+            handler(nullptr, std::make_exception_ptr(PermissionActionException(error_str, code)));
         }
         else {
-            callback({});
+            handler(&*object, nullptr);
```
Comment (bdash) [inline src/sync/sync_permission.cpp]: Isn't `&*` here equivalent to `object.get()`?

File: src/sync/sync_permission.cpp [hunk 8]
```diff
@@ -157,7 +157,85 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
...
+                               const ConfigMaker& make_config)
+{
+    perform_async_operation("PermissionOfferResponse",
+                            std::move(user),
+                            make_handler_extracting_property("realmUrl", std::move(callback)),
+                            AnyDict{ {"token", token} },
+                            make_config);
+}
+
+Permissions::AsyncOperationHandler Permissions::make_handler_extracting_property(std::string property,
```
Comment (bdash) [inline src/sync/sync_permission.cpp]: Is there any reason for this to be a static member function rather than just a static function in the .cpp file?

File: src/sync/sync_permission.cpp (hunk 7 above)
Comment (austinzheng) [reply src/sync/sync_permission.cpp]: Ah, that's it.

**Comment by austinzheng** [comment]
@bdash you are correct about the tests. We can't test anything dependent on ROS at the object store level currently.

**Comment by nirinchev** [comment]
> Rationale: right now, most bindings (Cocoa included) use the object store APIs to get and change permissions, but binding-level APIs to create and accept permission offers. 

To clarify, Cocoa is the only binding that uses the object store API for anything. .NET, Java, and js have convenience wrappers around the binding-level objects - none of them touches OS.
